aiida/orm/implementation/sqlalchemy/groups.py

```diff
@@ -161,40 +162,61 @@ def next(self):
 
         return Iterator(self._dbmodel.dbnodes, self._backend)
 
-    def add_nodes(self, nodes):
+    def add_nodes(self, nodes, **kwargs):
         """Add a node or a set of nodes to the group.
 
         :note: all the nodes *and* the group itself have to be stored.
 
         :param nodes: a list of `BackendNode` instance to be added to this group
+
+        :param kwargs:
+            skip_orm: When the flag is on, the SQLA ORM is skipped and SQLA is used
+            to create a direct SQL INSERT statement to the group-node relationship
+            table (to improve speed).
         """
         from sqlalchemy.exc import IntegrityError  # pylint: disable=import-error, no-name-in-module
+        from sqlalchemy.dialects.postgresql import insert  # pylint: disable=import-error, no-name-in-module
         from aiida.orm.implementation.sqlalchemy.nodes import SqlaNode
         from aiida.backends.sqlalchemy import get_scoped_session
+        from aiida.backends.sqlalchemy.models.base import Base
 
         super(SqlaGroup, self).add_nodes(nodes)
+        skip_orm = kwargs.get('skip_orm', False)
 
-        with utils.disable_expire_on_commit(get_scoped_session()) as session:
+        def check_node(given_node):
+            """ Check if given node is of correct type and stored """
+            if not isinstance(given_node, SqlaNode):
+                raise TypeError('invalid type {}, has to be {}'.format(type(given_node), SqlaNode))
+
+            if not given_node.is_stored:
+                raise ValueError('At least one of the provided nodes is unstored, stopping...')
 
+        with utils.disable_expire_on_commit(get_scoped_session()) as session:
             # Get dbnodes here ONCE, otherwise each call to dbnodes will re-read the current value in the database
             dbnodes = self._dbmodel.dbnodes
```

Comment:
This line is only used in the `not skip_orm` case, so could be moved in the if-block correct?

Reply:
I am not totally sure.

At https://docs.sqlalchemy.org/en/latest/core/connections.html I see that the connection.execute() statements (for inserts) are followed by commit() statements. Which implies that there should be a commit() after an execute() and if there is a commit(), there is a question of what happens with the objects that are in the session.

I wanted to follow a uniform approach and not expire the objects only in the default implementation and not in the code that optimised (I suppose that this will decrease the benefits of the optimized code).

The only case that I can imagine that could be problematic is that since we issue (indirectly) an SQL statement to add nodes to a group, then the group (at the SQLA-ORM level) might not be aware of this additions with the expire_on_commit=false. However, it is not the case since there is a test that covers this case

Reply:
Ah no, I meant the line `dbnodes = self._dbmodel.dbnodes` because `dbnodes` is only used in the first if-block

Reply:
Yes, you are right. I iterate on nodes and not on dbnodes, it can go even further down at the "not skip_orm" if-case.